File: Sources/SomnioServerCore/Bootstrap/RunServer.swift
import Foundation
import Logging
import PostgresNIO
import ServiceLifecycle
import SomnioCore
import SomnioData

/// Entry point the `SomnioServer` executable shim calls. Owns the entire bootstrap:
/// logging → config resolution → `PostgresClient` construction → `ServiceGroup` setup
/// (with `PostgresClient` as the initial service) → post-readiness sequence
/// (`waitForClientQueryable` → `MigrationRunner.applyPending` → `SectorCache.load` →
/// `WorldRouter` construction → world-clock pre-load → `WorldClockService` construction)
/// → append `Application`, `WorldRouter`, `CheckpointService`, `WorldClockService`,
/// `AITickService` → terminal `try await groupTask.value`.
///
/// `(environment, isDebug)` are injected so tests can drive the entry with a synthesized
/// environment without monkey-patching globals.
public func runServer(
    environment: [String: String] = ProcessInfo.processInfo.environment,
    isDebug: Bool = isDebugBuild
) async throws {
    ServerLoggingConfiguration.bootstrap()
    let lifecycleLogger = Logger(label: "de.tobiha.somnio.server.lifecycle")
    let postgresLogger = Logger(label: "de.tobiha.somnio.server.gameplay.persistence.postgres")
    // Migrations are an admin/operator concern (one-shot at boot), not gameplay traffic, so
    // the label routes through the admin file backend instead of `gameplay-log.log`.
    let migrationsLogger = Logger(label: "de.tobiha.somnio.server.admin.migrations")
    let sectorsLogger = Logger(label: "de.tobiha.somnio.server.gameplay.sectors.loader")

    let postgresConfiguration: PostgresClient.Configuration
    let serverConfiguration: ServerConfiguration
    do {
        postgresConfiguration = try resolvePostgresConfiguration(environment: environment, isDebug: isDebug)
        serverConfiguration = try ServerConfiguration.resolve(environment: environment, isDebug: isDebug)
    } catch {
        lifecycleLogger.error("failed to resolve configuration", metadata: ["error": "\(error)"])
        throw error
    }

    let client = PostgresClient(configuration: postgresConfiguration, backgroundLogger: postgresLogger)

    var groupConfiguration = ServiceGroupConfiguration(
        services: [
            ServiceGroupConfiguration.ServiceConfiguration(
                service: client,
                successTerminationBehavior: .gracefullyShutdownGroup,
                failureTerminationBehavior: .gracefullyShutdownGroup
            )
        ],
        gracefulShutdownSignals: [.sigint, .sigterm],
        logger: lifecycleLogger
    )
    // Cap total shutdown time so a hung Postgres snapshot during drain can't stall shutdown
    // indefinitely. The reverse-shutdown order still applies inside this cap.
    groupConfiguration.maximumGracefulShutdownDuration = .seconds(15)
    let group = ServiceGroup(configuration: groupConfiguration)

    let groupTask = Task { try await group.run() }

    let worldRouter: WorldRouter
    let worldClockService: WorldClockService
    let dependencies: ConnectionDependencies
    let adminDependencies: AdminConnectionDependencies
    do {
        try await waitForClientQueryable(client, logger: postgresLogger)
        try await MigrationRunner(client: client, logger: migrationsLogger).applyPending()
        let sectorCache = SectorCache()
        try await sectorCache.load(from: serverConfiguration.sectorsDirectory)
        let sectorNames = await sectorCache.names()
        sectorsLogger.info(
            "sector cache populated",
            metadata: ["count": "\(sectorNames.count)"]
        )
        try requireSectorsLoaded(sectorNames)
        let accounts = PostgresAccountRepository(client: client, logger: postgresLogger)
        let characters = PostgresCharacterRepository(client: client, logger: postgresLogger)
        let inventories = PostgresInventoryRepository(client: client, logger: postgresLogger)
        let registrations = PostgresRegistrationRepository(client: client, logger: postgresLogger)
        let npcDialogStates = PostgresNPCDialogStateRepository(client: client, logger: postgresLogger)
        let worldClocks = PostgresWorldClockRepository(client: client, logger: postgresLogger)
        let passwordHasher = PasswordHasher(logger: postgresLogger)
        let worldRouterLogger = Logger(label: "de.tobiha.somnio.server.gameplay.world")
        worldRouter = try await WorldRouter(
            sectors: sectorCache.snapshotByName(),
            characters: characters,
            npcDialogStates: npcDialogStates,
            logger: worldRouterLogger
        )
        // Pre-load the world clock synchronously on the readiness path so a login that
        // arrives in the first 100 ms after the service group reports ready sees the
        // persisted clock rather than the boot-default fallback.
        let initialClock = try await worldClocks.load()
        worldClockService = WorldClockService(
            worldRouter: worldRouter,
            worldClocks: worldClocks,
            initialClock: initialClock,
            logger: Logger(label: "de.tobiha.somnio.server.gameplay.worldclock")
        )
        dependencies = ConnectionDependencies(
            accounts: accounts,
            characters: characters,
            inventories: inventories,
            registrations: registrations,
            passwordHasher: passwordHasher,
            worldRouter: worldRouter,
            worldClock: worldClockService,
            configuration: serverConfiguration,
            logger: Logger(label: "de.tobiha.somnio.server.gameplay.connection")
        )
        adminDependencies = AdminConnectionDependencies(
            worldRouter: worldRouter as any AdminWorldRouter,
            worldClock: worldClockService,
            serverVersion: SomnioServerVersion.value,
            logsDirectory: LoggingPaths.logsDirectory,
            gameplayLogFileName: ServerLoggingConfiguration.gameplayLogFileName,
            adminLogFileName: ServerLoggingConfiguration.adminLogFileName,
            logger: Logger(label: "de.tobiha.somnio.server.admin.dispatch")
        )
    } catch {
        lifecycleLogger.error("startup failed; shutting down", metadata: ["error": "\(error)"])
        await group.triggerGracefulShutdown()
        _ = try? await groupTask.value
        throw error
    }

    let application = makeSomnioServerApplication(
        configuration: serverConfiguration,
        postgres: client,
        dependencies: dependencies,
        adminDependencies: adminDependencies
    )
    let checkpointService = CheckpointService(
        worldRouter: worldRouter,
        interval: serverConfiguration.checkpointInterval,
        logger: Logger(label: "de.tobiha.somnio.server.gameplay.checkpoint")
    )
    let aiTickService = AITickService(
        worldRouter: worldRouter,
        logger: Logger(label: "de.tobiha.somnio.server.gameplay.aitick")
    )

    // Reverse-shutdown order is the inverse of registration order:
    // `AITickService` stops first so no in-flight tick contends with the connection drain;
    // `WorldClockService` stops second so no `DateTick` races into closing outboxes (and its
    // final per-shutdown clock save lands before any subsequent service teardown);
    // `CheckpointService` stops third so no new periodic writes contend with the drain; then
    // `WorldRouter.run()` drains every logged-in connection; the Hummingbird app closes the
    // accept loop; `PostgresClient` tears down. WebSocket connections stay live during the
    // world drain so per-connection writer tasks can flush their outboxes before the accept
    // loop closes; the trade-off is that a login racing in mid-drain can be silently dropped
    // from `loggedInConnections.removeAll()`, but its own `ConnectionActor` cleanup still
    // snapshots through the disconnect path. `addServiceUnlessShutdown` silently no-ops if
    // the group is already shutting down, which is the actual safety net against a SIGTERM
    // during the readiness window.
    let postReadinessServices: [any Service] = [
        application,
        worldRouter,
        checkpointService,
        worldClockService,
        aiTickService
    ]
    for service in postReadinessServices {
        await group.addServiceUnlessShutdown(
            ServiceGroupConfiguration.ServiceConfiguration(
                service: service,
                successTerminationBehavior: .gracefullyShutdownGroup,
                failureTerminationBehavior: .gracefullyShutdownGroup
            )
        )
    }

    lifecycleLogger.info("SomnioServer ready; awaiting termination signal")

    try await groupTask.value
}

File: packages/server/src/bootstrap/runServer.ts
import { resolve } from 'node:path'
import { setTimeout as delay } from 'node:timers/promises'
import {
  PostgresAccountRepository,
  PostgresCharacterRepository,
  PostgresInventoryRepository,
  PostgresNPCDialogStateRepository,
  PostgresRegistrationRepository,
  PostgresSessionRepository,
  PostgresWorldClockRepository,
  assertQueryable,
  createDatabase,
  migrateToLatest,
  resolvePostgresConfiguration,
} from '@somnio/data'
import type { SomnioDatabase } from '@somnio/data'
import { resolveServerConfiguration } from '../config.ts'
import type { ConnectionDependencies } from '../connection/dependencies.ts'
import type { AdminDependencies } from '../handlers/adminDispatcher.ts'
import { createApp } from '../http/app.ts'
import { startServer } from '../http/server.ts'
import { PRODUCTION_LOG_MAX_ARCHIVES, PRODUCTION_LOG_MAX_BYTES, createLogging } from '../logging.ts'
import type { Logging } from '../logging.ts'
import { loadSectorCache, requireSectorsLoaded } from '../sectors/sectorCache.ts'
import { AITickService } from '../services/aiTickService.ts'
import { CheckpointService } from '../services/checkpointService.ts'
import { WorldClockService } from '../services/worldClockService.ts'
import { SERVER_VERSION } from '../version.ts'
import { WorldRouter } from '../world/worldRouter.ts'
import { pruneOrphanNPCDialogStates } from './orphanDialogPrune.ts'

/** The log directory is fixed relative to the working directory; the image's `WORKDIR` owns it. */
const LOGS_DIRECTORY = 'logs'
const SHUTDOWN_CAP_MS = 15_000

/** Test seams over the production boot: every optional field defaults to what `runServer` uses. */
export interface BootOptions {
  logging: Logging
  /** Overrides the configured port; `0` binds an ephemeral one. */
  port?: number
  serverVersion?: string
  worldClockIntervalMs?: number
  aiTickIntervalMs?: number
  checkpointIntervalMs?: number
}

export interface BootedServer {
  port: number
  db: SomnioDatabase
  worldRouter: WorldRouter
  worldClock: WorldClockService
  dependencies: ConnectionDependencies
  /** The reverse of the boot: services stop, connections drain, the server closes, the pool destroys. */
  shutdown(): Promise<void>
}

/**
 * The whole boot, in order: config → database → readiness → migrations → sectors → orphan-dialog
 * prune → router → synchronous world-clock preload → listen → services.
 */
export async function bootServer(
  env: Record<string, string | undefined>,
  options: BootOptions
): Promise<BootedServer> {
  const logging = options.logging
  const lifecycle = logging.serverLogger('lifecycle')
  const configuration = resolveServerConfiguration(env)
  const postgresConfiguration = resolvePostgresConfiguration(env, configuration.devDefaults)
  const db = createDatabase(postgresConfiguration, (error) =>
    lifecycle.warn({ error: String(error) }, 'postgres pool error')
  )
  const failStartup = async (error: unknown): Promise<never> => {
    lifecycle.error({ error: String(error) }, 'startup failed; shutting down')
    await db.destroy()
    throw error
  }
  try {
    await assertQueryable(db)
    await migrateToLatest(db)
  } catch (error) {
    return failStartup(error)
  }

  const sectorsLogger = logging.gameplayLogger('sectors.loader')
  let sectors: ReturnType<typeof loadSectorCache>
  try {
    sectors = loadSectorCache(configuration.sectorsDirectory)
    sectorsLogger.info({ count: sectors.size }, 'sector cache populated')
    requireSectorsLoaded(sectors, configuration.sectorsDirectory)
  } catch (error) {
    return failStartup(error)
  }

  const accounts = new PostgresAccountRepository(db)
  const characters = new PostgresCharacterRepository(db)
  const inventories = new PostgresInventoryRepository(db)
  const registrations = new PostgresRegistrationRepository(db)
  const npcDialogStates = new PostgresNPCDialogStateRepository(db)
  const worldClocks = new PostgresWorldClockRepository(db)
  const sessions = new PostgresSessionRepository(db)

  let worldRouter: WorldRouter
  let worldClock: WorldClockService
  try {
    // The prune runs before the router seeds from the table, so it never loads a cursor for an
    // NPC index the loaded sector no longer has.
    await pruneOrphanNPCDialogStates(
      npcDialogStates,
      sectors,
      configuration.forceDialogPrune,
      logging.adminLogger('dialogprune')
    )
    worldRouter = await WorldRouter.create(
      sectors,
      characters,
      npcDialogStates,
      logging.gameplayLogger('world'),
      logging.gameplayLogger('sector')
    )
    // Pre-loaded synchronously so a login in the first tick sees the persisted clock, not the default.
    worldClock = new WorldClockService(
      worldRouter,
      worldClocks,
      await worldClocks.load(),
      logging.gameplayLogger('worldclock'),
      options.worldClockIntervalMs
    )
  } catch (error) {
    return failStartup(error)
  }

  const dependencies: ConnectionDependencies = {
    accounts,
    characters,
    inventories,
    registrations,
    npcDialogStates,
    sessions,
    worldRouter,
    worldClock,
    outboxHighWatermark: configuration.outboxHighWatermark,
    logger: logging.gameplayLogger('connection'),
  }
  const adminDependencies: AdminDependencies = {
    worldRouter,
    worldClock,
    serverVersion: options.serverVersion ?? SERVER_VERSION,
    gameplayLog: logging.gameplayFile,
    adminLog: logging.adminFile,
    logger: logging.adminLogger('dispatch'),
  }
  let server: Awaited<ReturnType<typeof startServer>>
  try {
    server = await startServer({
      app: createApp({ db, healthLogger: logging.gameplayLogger('health') }),
      host: configuration.httpHost,
      port: options.port ?? configuration.httpPort,
      adminToken: configuration.adminToken,
      dependencies,
      adminDependencies,
      logger: logging.adminLogger('connection'),
    })
  } catch (error) {
    return failStartup(error)
  }

  const checkpoint = new CheckpointService(
    worldRouter,
    sessions,
    options.checkpointIntervalMs ?? configuration.checkpointIntervalMs,
    logging.gameplayLogger('checkpoint')
  )
  const aiTick = new AITickService(worldRouter, options.aiTickIntervalMs)
  const aiTickControl = new AbortController()
  const worldClockControl = new AbortController()
  const checkpointControl = new AbortController()
  const aiTickRun = aiTick.run(aiTickControl.signal)
  const worldClockRun = worldClock.run(worldClockControl.signal)
  const checkpointRun = checkpoint.run(checkpointControl.signal)
  lifecycle.info({ port: server.port, version: SERVER_VERSION }, 'SomnioServer ready')

  return {
    port: server.port,
    db,
    worldRouter,
    worldClock,
    dependencies,
    // The AI tick stops first so no in-flight tick contends with the drain; the world clock saves;
    // the checkpointer stops; the router drains every connection; then the server and the pool.
    shutdown: async () => {
      aiTickControl.abort()
      await aiTickRun
      lifecycle.debug('shutdown: ai tick stopped')
      worldClockControl.abort()
      await worldClockRun
      lifecycle.debug('shutdown: world clock saved')
      checkpointControl.abort()
      await checkpointRun
      lifecycle.debug('shutdown: checkpointer stopped')
      await worldRouter.drainAll()
      lifecycle.debug('shutdown: connections drained')
      await server.close()
      lifecycle.debug('shutdown: server closed')
      await db.destroy()
    },
  }
}

/** Boots, waits for SIGINT/SIGTERM, then shuts down under a 15 s cap. */
export async function runServer(env: Record<string, string | undefined>): Promise<void> {
  const logging = createLogging({
    directory: resolve(LOGS_DIRECTORY),
    maxBytes: PRODUCTION_LOG_MAX_BYTES,
    maxArchives: PRODUCTION_LOG_MAX_ARCHIVES,
  })
  const lifecycle = logging.serverLogger('lifecycle')
  const server = await bootServer(env, { logging })
  lifecycle.info('awaiting termination signal')

  await new Promise<void>((resolveSignal) => {
    const onSignal = (signal: NodeJS.Signals) => {
      lifecycle.info({ signal }, 'termination signal received')
      resolveSignal()
    }
    process.once('SIGINT', onSignal)
    process.once('SIGTERM', onSignal)
  })

  await Promise.race([server.shutdown(), delay(SHUTDOWN_CAP_MS, undefined, { ref: false })])
  lifecycle.info('SomnioServer stopped')
}
